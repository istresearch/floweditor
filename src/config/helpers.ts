import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  VISIBILITY_ONLINE
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (operators: Operator[], config: FlowEditorConfig): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (typeConfigs: Type[], config: FlowEditorConfig): Type[] => {
  return filterVisibility(typeConfigs, config);
};

const filterVisibility = (items: FlowTypeVisibility[], config: FlowEditorConfig): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (!(config.filters || []).find((name: string) => name === item.filter)) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1;
  });
};
