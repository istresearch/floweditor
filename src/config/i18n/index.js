import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

// TODO: consider chunking out translation files
// and using something like XHR backend to lazily
// load only the required language.
// See https://www.i18next.com/how-to/add-or-load-translations

import defaults from './defaults.json';
import cs from './cs/resource.json';
import es from './es/resource.json';
import fr from './fr/resource.json';
import mn from './mn/resource.json';
import ptBR from './pt-br/resource.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    detection: {
      order: ['querystring', 'htmlTag']
    },
    resources: {
      defaults: { translation: defaults },
      cs: { translation: cs },
      es: { translation: es },
      fr: { translation: fr },
      mn: { translation: mn },
      'pt-br': { translation: ptBR }
    },
    fallbackLng: ['defaults'],
    interpolation: {
      prefix: '[[',
      suffix: ']]'
    },
    lowerCaseLng: true
  });

export default i18n;
