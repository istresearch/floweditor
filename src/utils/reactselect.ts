import { StylesConfig } from 'react-select';

export const large = {
  placeholder: (styles: StylesConfig, state: any) => {
    return { ...styles, color: '#ddd', marginLeft: '3px' };
  },
  singleValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '3px' };
  },
  multiValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '0px' };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none',
      fontSize: '13px'
    };
  },
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      fontSize: '13px',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  }
};

export const messageStyle = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      borderBottom: 'none',
      borderBottomRightRadius: '0px',
      borderBottomLeftRadius: '0px',
      '&:hover': {
        borderColor: '#e6e6e6'
      }
    };
  }
};

export const getErroredSelect = (baseControl: any): any => {
  return {
    control: (styles: StylesConfig) => ({
      ...styles,
      ...baseControl,
      borderColor: 'tomato',
      boxShadow: '0 0 0 3px rgba(255,196,186,0.5) !important;',
      '&:hover': {
        borderColor: 'tomato'
      }
    })
  };
};

export const small = {
  control: (styles: StylesConfig) => ({
    ...styles,
    borderColor: '#e6e6e6',
    boxShadow: 'none',
    '&:hover': {
      borderColor: '#e6e6e6'
    },
    maxHeight: '28px',
    minHeight: '28px',
    fontSize: '12px',
    lineHeight: '12px',
    marginBottom: '-10px',
    '&:focus-within': {
      boxShadow: 'var(--widget-box-shadow-focused) !important',
      border: '1px solid #a4cafe !important'
    }
  }),
  menuList: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 200
  }),
  option: (styles: StylesConfig) => ({
    ...styles,
    fontSize: 12
  }),
  singleValue: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 28,
    lineHeight: '13px'
  }),
  indicatorsContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  valueContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  input: (styles: StylesConfig) => ({
    ...styles,
    marginLeft: '-2px',
    marginTop: '0px',
    caretColor: '#999',
    boxShadow: 'none',
    border: 'none',
    maxHeight: '20px',
    minHeight: '20px',
    lineHeight: '12px'
  })
};

export const tagging = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  multiValue: (styles: StylesConfig) => {
    return {
      ...styles,
      maxHeight: 24,
      minHeight: 24,
      lineHeight: '17px',
      backgroundColor: 'hsl(0, 0%, 94%)',
      border: '1px solid hsl(0, 0%, 85%)'
    };
  }
};
