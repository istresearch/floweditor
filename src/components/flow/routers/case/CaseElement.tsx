import { default as bindCallbacks } from 'auto-bind/react';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { isRelativeDate } from 'components/flow/routers/helpers';
import FormElement from 'components/form/FormElement';
import TextInputElement, { TextInputStyle } from 'components/form/textinput/TextInputElement';
import { fakePropType } from 'config/ConfigProvider';
import { filterOperators } from 'config/helpers';
import { Operator, Operators } from 'config/interfaces';
import { operatorConfigList } from 'config/operatorConfigs';
import { Case } from 'flowTypes';
import * as React from 'react';
import { FormState, StringEntry, SelectOptionEntry } from 'store/nodeEditor';
import { hasErrorType } from 'utils';

import styles from './CaseElement.module.scss';
import { initializeForm, validateCase } from './helpers';
import SelectElement, { SelectOption } from 'components/form/select/SelectElement';
import i18n from 'config/i18n';
import TembaSelect, { TembaSelectStyle } from 'temba/TembaSelect';

export interface CaseElementProps {
  kase: Case;
  categoryName: string;
  name?: string; // satisfy form widget props
  onRemove?(uuid: string): void;
  onChange?(c: CaseProps): void;
  operators?: Operator[];
  classifier?: any;
}

export interface CaseElementState extends FormState {
  errors: string[];
  operatorConfig: Operator;
  categoryName: StringEntry;
  categoryNameEdited: boolean;

  // for string based args
  argument: StringEntry;

  // for numeric operators
  min: StringEntry;
  max: StringEntry;

  // intents
  intent: SelectOptionEntry;
  confidence: StringEntry;

  state: StringEntry;
  district: StringEntry;
}

export default class CaseElement extends React.Component<CaseElementProps, CaseElementState> {
  private operators: Operator[];

  constructor(props: CaseElementProps) {
    super(props);

    bindCallbacks(this, {
      include: [/^handle/, /^get/]
    });

    this.state = initializeForm(props);
  }

  public static contextTypes = {
    config: fakePropType
  };

  public componentDidMount() {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      argument: this.state.argument.value,
      min: this.state.min.value,
      max: this.state.max.value,
      intent: this.state.intent.value,
      confidence: this.state.confidence.value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited,
      classifier: this.props.classifier
    });

    this.setState(updates as CaseElementState, this.handleChange);
  }

  public componentDidUpdate(previousProps: CaseElementProps): void {
    if (
      this.props.classifier &&
      this.props.classifier !== previousProps.classifier &&
      this.state.intent.value
    ) {
      const updates = validateCase({
        operatorConfig: this.state.operatorConfig,
        argument: this.state.argument.value,
        min: this.state.min.value,
        max: this.state.max.value,
        intent: this.state.intent.value,
        confidence: this.state.confidence.value,
        exitName: this.state.categoryName.value,
        exitEdited: this.state.categoryNameEdited,
        classifier: this.props.classifier
      });

      this.setState(updates as CaseElementState, this.handleChange);
    }
  }

  private getOperators(): Operator[] {
    let operators = this.props.operators || operatorConfigList;
    if (this.operators === undefined) {
      this.operators = filterOperators(operators, this.context.config);
    }

    return this.operators;
  }

  private getArgumentArray(): string[] {
    if (this.state.operatorConfig.operands === 0) {
      return [];
    }

    if (
      this.state.operatorConfig.type === Operators.has_intent ||
      this.state.operatorConfig.type === Operators.has_top_intent
    ) {
      if (this.state.intent.value) {
        return [this.state.intent.value.value, this.state.confidence.value];
      } else {
        return ['', this.state.confidence.value];
      }
    }

    if (this.state.operatorConfig.type === Operators.has_number_between) {
      return [this.state.min.value, this.state.max.value];
    }

    if (this.state.operatorConfig.type === Operators.has_ward) {
      return [this.state.state.value, this.state.district.value];
    }

    return [this.state.argument.value];
  }

  private handleOperatorChanged(operatorConfig: Operator): void {
    const updates = validateCase({
      operatorConfig,
      argument: this.state.argument.value,
      min: this.state.min.value,
      max: this.state.max.value,
      intent: this.state.intent.value,
      confidence: this.state.confidence.value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited,
      classifier: this.props.classifier
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleArgumentChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      argument: value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleDistrictChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      argument: this.state.argument.value,
      state: this.state.state.value,
      district: value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  /** The user changed the value for the state (a location-based state) */
  private handleStateChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      argument: this.state.argument.value,
      district: this.state.district.value,
      state: value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleIntentChanged(selected: SelectOption): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      intent: selected,
      confidence: this.state.confidence.value || '.9',
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited,
      classifier: this.props.classifier
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleConfidenceChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      intent: this.state.intent.value,
      confidence: value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited,
      classifier: this.props.classifier
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleMinChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      min: value,
      max: this.state.max.value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleMaxChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      min: this.state.min.value,
      max: value,
      exitName: this.state.categoryName.value,
      exitEdited: this.state.categoryNameEdited
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleExitChanged(value: string): void {
    const updates = validateCase({
      operatorConfig: this.state.operatorConfig,
      state: this.state.state.value,
      district: this.state.district.value,
      argument: this.state.argument.value,
      min: this.state.min.value,
      max: this.state.max.value,
      intent: this.state.intent.value,
      confidence: this.state.confidence.value,
      classifier: this.props.classifier,
      exitName: value,
      exitEdited: true
    });

    this.setState(updates as CaseElementState, () => this.handleChange());
  }

  private handleRemoveClicked(): void {
    this.props.onRemove(this.props.kase.uuid);
  }

  private getCaseProps(): CaseProps {
    const props = {
      uuid: this.props.kase.uuid,
      categoryName: this.state.categoryName.value,
      kase: {
        arguments: this.getArgumentArray(),
        type: this.state.operatorConfig.type,
        uuid: this.props.kase.uuid,

        // if the exit name changed, we'll need to recompute our exit
        category_uuid: this.state.categoryNameEdited ? null : this.props.kase.category_uuid
      },
      valid: this.state.valid
    };

    return props;
  }

  private handleChange(): void {
    // If the case doesn't have arguments & an exit name, remove it
    if (!this.state.categoryName.value) {
      // see if we are clearing out a between
      if (this.state.operatorConfig.type === Operators.has_number_between) {
        if (!this.state.min.value && !this.state.max.value) {
          // this.handleRemoveClicked();
          // return;
        }
      }
      // see if we are clearing out a single operand
      else {
        if (!this.state.argument.value) {
          // this.handleRemoveClicked();
          // return;
        }
      }
    }

    this.props.onChange(this.getCaseProps());
  }

  private handleIntentMenuOpened() {
    // hiding any errors when the menu opens
    this.setState({ intent: { value: this.state.intent.value } });
  }

  private handleIntentMenuClosed() {
    // we want to revalidate close without selection
    // wait a cycle for selection event to fire first
    window.setTimeout(() => {
      this.handleIntentChanged(this.state.intent.value);
    }, 0);
  }

  private renderArguments(): JSX.Element {
    if (this.state.operatorConfig.operands > 0) {
      // First pass at displaying, handling Operators.has_number_between inputs
      if (this.state.operatorConfig.operands > 1) {
        if (this.state.operatorConfig.type === Operators.has_number_between) {
          return (
            <>
              <TextInputElement
                name={i18n.t('forms.arguments', 'arguments')}
                style={TextInputStyle.small}
                onChange={this.handleMinChanged}
                entry={this.state.min}
                autocomplete={true}
              />
              <span className={styles.divider} data-draggable={true}>
                {i18n.t('forms.and', 'and')}
              </span>
              <TextInputElement
                name={i18n.t('forms.arguments', 'arguments')}
                style={TextInputStyle.small}
                onChange={this.handleMaxChanged}
                entry={this.state.max}
                autocomplete={true}
              />
            </>
          );
        } else if (
          this.state.operatorConfig.type === Operators.has_intent ||
          this.state.operatorConfig.type === Operators.has_top_intent
        ) {
          let intents: SelectOption[] = [];

          if (this.props.classifier && this.props.classifier.intents) {
            intents = this.props.classifier.intents.map((intent: string) => {
              const option: SelectOption = {
                name: intent,
                value: intent
              };
              return option;
            });
          }

          return (
            <>
              <div style={{ width: '114px' }}>
                <SelectElement
                  key="intent_select"
                  style={TembaSelectStyle.small}
                  name={i18n.t('forms.intent', 'Intent')}
                  placeholder={i18n.t('forms.select_intent', 'Select intent')}
                  entry={this.state.intent}
                  onChange={this.handleIntentChanged}
                  options={intents}
                  onMenuOpen={this.handleIntentMenuOpened}
                  onMenuClose={this.handleIntentMenuClosed}
                  hideError={true}
                ></SelectElement>
              </div>
              <div className={styles.divider} data-draggable={true}>
                above
              </div>
              <div style={{ width: '34px' }}>
                <TextInputElement
                  name={i18n.t('forms.confidence', 'confidence')}
                  onChange={this.handleConfidenceChanged}
                  entry={this.state.confidence}
                  style={TextInputStyle.small}
                  placeholder=".9"
                />
              </div>
            </>
          );
        } else {
          return (
            <>
              <TextInputElement
                name={i18n.t('forms.state', 'State')}
                placeholder="State"
                onChange={this.handleStateChanged}
                style={TextInputStyle.small}
                entry={this.state.state}
              />
              <span className={styles.divider} data-draggable={true}>
                and
              </span>
              <TextInputElement
                name={i18n.t('forms.district', 'District')}
                placeholder={i18n.t('forms.district', 'District')}
                onChange={this.handleDistrictChanged}
                style={TextInputStyle.small}
                entry={this.state.district}
              />
            </>
          );
        }
      } else if (isRelativeDate(this.state.operatorConfig.type)) {
        return (
          <>
            <span className={styles.divider} data-draggable={true}>
              today +{' '}
            </span>
            <TextInputElement
              __className={styles.relative_date}
              name={i18n.t('forms.arguments', 'arguments')}
              onChange={this.handleArgumentChanged}
              entry={this.state.argument}
              style={TextInputStyle.small}
              autocomplete={false}
            />
            <span className={styles.divider}>days</span>
          </>
        );
      } else {
        return (
          <TextInputElement
            data-test-id="case-arguments"
            name={i18n.t('forms.arguments', 'arguments')}
            onChange={this.handleArgumentChanged}
            entry={this.state.argument}
            style={TextInputStyle.small}
            placeholder={this.state.operatorConfig.type === Operators.has_district ? 'State' : ''}
            autocomplete={true}
          />
        );
      }
    }

    return null;
  }

  public render(): JSX.Element {
    return (
      <FormElement
        data-spec="case-form"
        name={this.props.name}
        __className={styles.group}
        kaseError={this.state.errors.length > 0}
      >
        <div
          className={`${styles.kase} ${styles[this.state.operatorConfig.type]}`}
          data-draggable={true}
        >
          <span className={`fe-chevrons-expand ${styles.dnd_icon}`} data-draggable={true} />
          <div className={styles.choice}>
            <TembaSelect
              name={i18n.t('forms.operator', 'operator')}
              style={TembaSelectStyle.small}
              options={this.getOperators()}
              nameKey="verboseName"
              valueKey="type"
              onChange={this.handleOperatorChanged}
              value={this.state.operatorConfig}
            ></TembaSelect>
          </div>
          <div
            className={
              this.state.operatorConfig.operands > 1 ? styles.multi_operand : styles.single_operand
            }
          >
            {this.renderArguments()}
          </div>
          <div className={styles.categorize_as} data-draggable={true}>
            {i18n.t('forms.categorize_as', 'categorize as')}
          </div>
          <div className={styles.category}>
            <TextInputElement
              name={i18n.t('forms.exit_name', 'Exit Name')}
              style={TextInputStyle.small}
              onChange={this.handleExitChanged}
              entry={this.state.categoryName}
              maxLength={36}
              showInvalid={hasErrorType(this.state.errors, [/category/])}
            />
          </div>
          <span
            data-testid={'remove-case-' + this.props.kase.uuid}
            className={`fe-x ${styles.remove_icon}`}
            onClick={this.handleRemoveClicked}
          />
        </div>
      </FormElement>
    );
  }
}
