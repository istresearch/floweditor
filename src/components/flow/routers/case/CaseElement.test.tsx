import CaseElement, { CaseElementProps } from 'components/flow/routers/case/CaseElement';
import { Operators } from 'config/interfaces';
import * as React from 'react';
import { fireEvent, render, fireTembaSelect } from 'test/utils';
import { createUUID } from 'utils';

const caseUUID = createUUID();

const caseProps: CaseElementProps = {
  name: `case_${caseUUID}`,
  kase: {
    uuid: caseUUID,
    type: Operators.has_any_word,
    arguments: [''],
    category_uuid: createUUID()
  },
  categoryName: null,
  onRemove: jest.fn(),
  onChange: jest.fn()
};

const selectOperator = (getByTestId: any, operator: Operators) => {
  fireTembaSelect(getByTestId('temba_select_operator'), [{ type: operator }]);
};

describe(CaseElement.name, () => {
  describe('render', () => {
    it('should render empty case', () => {
      const { baseElement } = render(<CaseElement {...caseProps} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('renders no argument rules', () => {
      const kase = {
        uuid: caseUUID,
        type: Operators.has_number,
        category_uuid: createUUID()
      };

      const { baseElement } = render(<CaseElement {...caseProps} kase={kase} />);
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('operator changes', () => {
    // we need a full renders because we update our category reference
    it('should handle updates', () => {
      const { baseElement, getByTestId } = render(<CaseElement {...caseProps} />);

      selectOperator(getByTestId, Operators.has_phone);
      expect(baseElement).toMatchSnapshot();
    });

    it('should should set arguments for numeric range', () => {
      const { baseElement, getByTestId } = render(<CaseElement {...caseProps} />);
      selectOperator(getByTestId, Operators.has_number_between);
      expect(baseElement).toMatchSnapshot();
    });

    it('shouldnt update exit if it has been edited', () => {
      const { baseElement, queryByDisplayValue, getByTestId, getByDisplayValue } = render(
        <CaseElement {...caseProps} />
      );

      // make us a has phone so we can look up our category by value
      selectOperator(getByTestId, Operators.has_phone);

      // update our category to a user supplied value
      const category = getByDisplayValue('Has Phone');
      fireEvent.change(category, { target: { value: 'My Exit Name' } });

      // now swithc our type to force a category change
      selectOperator(getByTestId, Operators.has_number);

      // we shouldn't have updated our category
      expect(queryByDisplayValue('My Exit Name')).not.toBeNull();
      expect(queryByDisplayValue('Has Number')).toBeNull();
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('handles removes', () => {
      const onRemove = jest.fn();
      const { getByTestId } = render(<CaseElement {...caseProps} onRemove={onRemove} />);

      fireEvent.click(getByTestId(`remove-case-${caseProps.kase.uuid}`));
      expect(onRemove).toHaveBeenCalled();
    });

    it('handles argument change', () => {
      const onRemove = jest.fn();
      const { baseElement, getAllByTestId } = render(
        <CaseElement {...caseProps} onRemove={onRemove} />
      );
      const args = getAllByTestId('arguments');
      fireEvent.change(args[0], { target: { value: 'Purple, p' } });
      expect(baseElement).toMatchSnapshot();
    });

    it('handles multiple argument change', () => {
      const onRemove = jest.fn();
      const { baseElement, getByTestId, getAllByTestId } = render(
        <CaseElement {...caseProps} onRemove={onRemove} />
      );

      selectOperator(getByTestId, Operators.has_number_between);

      const args = getAllByTestId('arguments');
      fireEvent.change(args[0], { target: { value: '1' } });
      fireEvent.change(args[1], { target: { value: '100' } });
      expect(baseElement).toMatchSnapshot();
    });
  });
});
