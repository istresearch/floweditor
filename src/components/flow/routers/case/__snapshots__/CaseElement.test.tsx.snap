// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseElement operator changes should handle updates 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_phone"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_phone\\",\\"verboseName\\":\\"has a phone number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Phone\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value="Has Phone"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement operator changes should should set arguments for numeric range 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number_between"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number_between\\",\\"verboseName\\":\\"has a number between\\",\\"operands\\":2}]"
            />
          </div>
        </div>
        <div
          class="multi_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                value=""
              />
            </div>
          </div>
          <span
            class="divider"
            data-draggable="true"
          >
            and
          </span>
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value=""
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement operator changes shouldnt update exit if it has been edited 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number\\",\\"verboseName\\":\\"has a number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Number\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value="My Exit Name"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement render renders no argument rules 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number\\",\\"verboseName\\":\\"has a number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Number\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value="Has Number"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement render should render empty case 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_any_word"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_any_word\\",\\"verboseName\\":\\"has any of the words\\",\\"operands\\":1}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                placeholder=""
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value=""
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement update handles argument change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_any_word"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_any_word\\",\\"verboseName\\":\\"has any of the words\\",\\"operands\\":1}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                placeholder=""
                value="Purple, p"
              />
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value="Purple"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement update handles multiple argument change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number_between"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              expressions=""
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number_between\\",\\"verboseName\\":\\"has a number between\\",\\"operands\\":2}]"
            />
          </div>
        </div>
        <div
          class="multi_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                value="1"
              />
            </div>
          </div>
          <span
            class="divider"
            data-draggable="true"
          >
            and
          </span>
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="arguments"
                name="arguments"
                value="100"
              />
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Exit Name"
                name="Exit Name"
                value="1 - 100"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;
